src/dc_tools.rs

```diff
@@ -1256,37 +1248,34 @@ pub unsafe fn dc_write_file(
     buf: *const libc::c_void,
     buf_bytes: size_t,
 ) -> libc::c_int {
-    let mut success = 0;
-    let pathNfilename_abs = dc_get_abs_path(context, pathNfilename);
+    let bytes = std::slice::from_raw_parts(buf as *const u8, buf_bytes);
 
+    dc_write_file_safe(context, as_str(pathNfilename), bytes) as libc::c_int
+}
+
+pub fn dc_write_file_safe(context: &Context, pathNfilename: impl AsRef<str>, buf: &[u8]) -> bool {
```

Comment:
If you actually wanted a safe one it'd take `PathBuf` as parameter no?  But this refactoring is all out of scope anyway i think